PST Level up Proficiencies w/table & fixes

chilvence opened the pull request:
This is a redo of #1023, implementing the individual character proficiencies via a 2da table

Also includes the various changes to adjacent code : fixing the specific checks, shuffling the hp/thaco, slight tidy up of saving throws

If this is all ok, I can easily move on to finish #176 , all that is remaining are spell slots (which trivially works with LUSpellSelection) and the specialisation bonuses for TNO

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] I used the same coding style as the surrounding code <!-- yet to be formally defined, see issue #161 -->
- [ ] I have tested the proposed changes
- [ ] I extended the documentation, if necessary
- [ ] The proposed change builds also on our build bots (check after submission)

lynxlynxlynx commented on gemrb/GUIScripts/pst/GUIREC.py:
```diff
@@ -1191,17 +1192,22 @@ def GetThac0 (Class, Level):
+		return maxProf - currProf
+
+	# Nameless continues gaining points forever	at a rate of 1 every 3 levels
+	elif AvatarName == "NAMELESS" and (currProf-3) <= (nextLevel / 3):
```
This should be a function instead: `AvatarName == "NAMELESS"`
It can also replace the `Specific == 2` checks.

chilvence replied:
Good idea, should make the whole thing more readable

lynxlynxlynx commented:
Some CurrWeapProf use looks weird, but if you say it works, that's good enough for me, considering the current state.

chilvence commented:
I was thinking of renaming that variable to something like CurrentTotal, it's just meant to be the total number of points the character has already across all proficiencies

How do you mean by weird though? Honest question, I've been staring at this too long so I can't really see it

lynxlynxlynx commented on gemrb/GUIScripts/pst/GUIREC.py:
```diff
@@ -826,7 +826,10 @@ def AcceptLevelUp():
 	#increase weapon proficiency if needed
 	if WeapProfType!=-1:
-		GemRB.SetPlayerStat (pc, WeapProfType, CurrWeapProf + WeapProfGained )
+		GemRB.SetPlayerStat (pc, IE_PROFICIENCYBASTARDSWORD+WeapProfType, CurrWeapProf + WeapProfGained )
```
This one is the puzzling one. Aren't you setting one specific prof to the sum of all profs?

chilvence replied:
Oh I see - yes that is exactly right. Each character only trains in one stat, and it all happens automatically. Otherwise I would have just plugged the whole thing into LUProfsSelection

TNO is the exception, he gains the points in the last proficiency slot, and then dialog scripts with npcs are used to actually train specific weapons (which I am pretty sure already works last time I checked)

I can see how that could all use some explanatory commenting though, it's all a bit cryptic.

lynxlynxlynx commented:
The build failure is unrelated, but here's the linter output for convenience:
```
 Successfully installed pycodestyle-2.6.0
./gemrb/GUIScripts/pst/GUIREC.py:825:17: E225 missing whitespace around operator
	if WeapProfType!=-1:
	               ^
./gemrb/GUIScripts/pst/GUIREC.py:827:98: E202 whitespace before ')'
		GemRB.SetPlayerStat (pc, IE_PROFICIENCYBASTARDSWORD+WeapProfType, CurrWeapProf + WeapProfGained )
		                                                                                               ^
./gemrb/GUIScripts/pst/GUIREC.py:831:68: E202 whitespace before ')'
		GemRB.SetPlayerStat (pc, IE_FREESLOTS, freeSlots + WeapProfGained )
		                                                                 ^
```

chilvence commented:
I squashed the commits around that and fixed the whitespace

I don't really know the best way of checking style issues like that, I wouldn't mind checking but I don't know what tool to use.

lynxlynxlynx commented:
In this case it's pycodestyle (as quoted) and here you can see roughly how to invoke it:
https://github.com/gemrb/gemrb/blob/d03dff74831914a2bea7531bf0128a4ef462d762/.github/workflows/style.yml#L29

chilvence commented:
Ah nice, I'll do that in future to avoid flooding everyone's inbox with error notifications (sorry...)

lynxlynxlynx commented:
You have it here otherwise:
![image](https://user-images.githubusercontent.com/121515/99056742-093bba80-259b-11eb-8362-46ad3b06e1d1.png)

chilvence commented:
Whoops. 9th time's the charm...

lynxlynxlynx commented:
Thanks!
